core: refactor and improve surface commit

ikalco opened the pull request:
#### Describe your PR, what does it fix/add?
part 2 of #9774

relevant commits:
[cleanup SSurfaceState and surface pending commit tracking](https://github.com/hyprwm/Hyprland/pull/9805/commits/dfcfb1d68306ceb73215c14536390a63070a553d)
[move surface code from DRMSyncobj, and move acquire to SSurfaceState](https://github.com/hyprwm/Hyprland/pull/9805/commits/9fbd34d0c75d4b3f76954300520164cb4999b9e0)
[use queue for comitted pending surface states like proto says](https://github.com/hyprwm/Hyprland/pull/9805/commits/3cb43e04b1b83fb81ebc2b78f1572a3c54f07c68)
[drop, not release, prev buffer if 2nd buffer wl_surface.attach is sent](https://github.com/hyprwm/Hyprland/pull/9805/commits/3002845e0e384f89fe872b9325142ee1d43a6230)

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
based on part 1

#### Is it ready for merging, or does it need work?
<s>should be gtg now</s>
<s>nvm lol nvidia borkde again (╥_╥)</s>
yes finally

gulafaran commented:
currently discord typing is occasionally delayed, not updating. minecraft flickers, cs2 flickers.

gulafaran commented:
and obs -> right click preview -> open windowed projector should die with 

```
[ 653985.111] {Display Queue} wl_display#1.error(wp_linux_drm_syncobj_surface_v1#77, 4, "Missing acquire timeline")
wp_linux_drm_syncobj_surface_v1#77: error 4: Missing acquire timeline
```

as it does on -git, currently doesnt.

ikalco commented:
> and obs -> right click preview -> open windowed projector should die with
> 
> ```
> [ 653985.111] {Display Queue} wl_display#1.error(wp_linux_drm_syncobj_surface_v1#77, 4, "Missing acquire timeline")
> wp_linux_drm_syncobj_surface_v1#77: error 4: Missing acquire timeline
> ```
> 
> as it does on -git, currently doesnt.

fixed

ikalco commented:
works on nvidia now, no flicker or delayed frames :)
I had to dig up my old nvidia card, and thank god I could repro lol

PlasmaPower commented:
Did you test this with the vkcube FIFO mode Tom mentioned was broken? That might've been from another PR but still worth testing here

ikalco commented:
tested with explicit_sync both on and off, all work on nvidia
* minecraft, vsync and unlimited, windowed and ds
* csgo, windowed and ds
* obs recording
* `vkcube --wsi wayland --present_mode {number here}` with 0, 1, and 2

ikalco commented:
@vaxerski do you mind if we just merge this as one pr lol, cause I don't really wanna go through all the combinations of testing all over again for #9804

maybe we can do something like
git squash the buffer ref stuff into one commit
git squash the surface stuff into another
and merge those 2 kernel style like you said before

p.s. the other two prs aren't required for this, still drafts until I get the time to clean/fix them up and test

PlasmaPower commented:
I've been running this PR for a couple days and it seems to be working well for my usecases on NVIDIA.

ikalco commented:
rebased to main

vaxerski commented:
my question is what is this supposed to fix?

ikalco commented:
> my question is what is this supposed to fix?

there isn't a specific bug this pr is fixing, but it is building towards more well defined surface commit code
so that in something like my proper buffer release pr (#9807) I know what gets called where,
and can hopefully fix the last of those pesky nvidia flickering bugs (#9640)

this pr also follows compositor protocol closer, like in the last 2 commits
and while I was in there changing all of that, I also decided to clean up SSurfaceState by
* making `updated` easier to use
* setting the right initial values based on the protocol text
* grouping functions and adding some clarifying comments

----

p.s.
<details><summary>here are some exmaples of previously not well defined behavior</summary>

* previously, buffer refs were tracked using a `SP<CHLBufferReference>` which has 2 ref counting methods, the SP and buffer->nLocks... that's confusing and unneeded, especially for having proper buffer releases
* release and acquire points were held in the CHLBufferReference, [which is wrong](https://github.com/hyprwm/Hyprland/pull/9774#discussion_r2019681644)
  this pr moves the acquire point to SurfaceState,
  and the release will be done properly in the later release stuff pr
* some commitPendingState calls were obscured through a events.precommit.emit() and done in DRMSyncobj protocol
  which again adds some confusion as to how and where state is committed
  and sometimes the DRMSyncobj protocol committed state incorrectly
  ```cpp
  if (!surface->pending.buffer && !PENDING_HAS_NEW_BUFFER && surface->current.buffer) {
          surface->current.bufferDamage.clear();
          surface->current.damage.clear();
          surface->commitPendingState(surface->current);
          return; // no new buffer, but we still have current around and a commit happend, commit current again.
    }
  ```

* previously, the pending surface state after a commit was being reset in multiple places in different ways, now we just have SSurfaceState::reset() to clearly define that new pending state
* simlarly to right above, the SSurfaceState::updateFrom just copied everything when a buffer commit happened, which is wrong. Now we only copy committed things
  ```cpp
  if (ref.updated & SURFACE_UPDATED_BUFFER) {
          ref.updated &= ~SURFACE_UPDATED_BUFFER;
          *this = ref;
          ref.damage.clear();
          ref.bufferDamage.clear();
          ref.buffer = {};
  }
  ```

</details>

vaxerski commented on src/protocols/types/SurfaceState.hpp:
```diff
@@ -2,44 +2,59 @@
+    union {
+        uint16_t all = 0;
+        struct {
+            bool buffer : 1;
```
first time I see this syntax what the fuck? xD

ikalco replied:
each bool takes up one bit in the 2byte union and compiler does all the bit shifting stuff for us so its nice to use
I can make it just a struct of bools if you want, but I bet the compiler will just optimize it like above

PlasmaPower replied:
If you made a struct of bools normally I think the compiler would have to put each in its own byte for alignment and the ability to take a pointer to each of them

ikalco replied:
maybe  ¯\_(ツ)_/¯, who knows its the cpp compiler

ikalco replied:
https://www.programiz.com/online-compiler/65xnymqumVjNH

`dothething()` basically ANDs `a` and `b`, so only true if both `a` and `b` are true
and it only works in that scenario :+1: 

also, heres a reference for the syntax, its called a [bit-field](https://en.cppreference.com/w/cpp/language/bit_field)

vaxerski replied:
huh. I know what a bitfield is, just never seen this syntax. Great, guess we all learn every day. Thanks!

vaxerski commented on src/protocols/types/SurfaceState.cpp:
```diff
@@ -45,49 +45,56 @@ void SSurfaceState::updateSynchronousTexture(SP<CTexture> lastTexture) {
+    updated.all = false;
+
+    // After commit, there is no pending buffer until the next attach.
+    buffer = {};
```
who said so? I've seen apps that do:
```
attach buffer
commit
set scale 
set something
commit
attach buffer 2 
commit
...
```

ikalco replied:
https://wayland.app/protocols/wayland#wl_surface:request:attach
>The initial surface contents are void; there is no content. wl_surface.attach assigns the given wl_buffer as the pending wl_buffer. wl_surface.commit makes the pending wl_buffer the new surface contents, and the size of the surface becomes the size calculated from the wl_buffer, as described above. **After commit, there is no pending buffer until the next attach.**

vaxerski replied:
oke, I cant read, am a zoomer

vaxerski commented:
cs works fine on my end so yea
